rectools/visuals/metrics_app.py

```diff
@@ -0,0 +1,212 @@
+
+    @property
+    def metric_names(self) -> tp.List[str]:
+        """List of metric column names from the `models_metrics`."""
+        non_metric_columns = {Columns.Split, Columns.Model}
+        return [col for col in self.models_metrics.columns if col not in non_metric_columns]
+
+    @property
+    def model_names(self) -> tp.List[str]:
+        """Sorted list of model names from `models_metrics`."""
+        return sorted(self.models_metrics[Columns.Model].unique())
+
+    @property
+    def fold_ids(self) -> tp.List[int]:
+        """Sorted list of fold identifiers from the `models_metrics`."""
+        return sorted(self.models_metrics[Columns.Split].unique())
+
+    @property
+    def n_folds(self) -> int:
+        """Total number of unique folds available in the `models_metrics`."""
+        return len(self.fold_ids)
+
+    def _validate_input_data(self) -> None:
+        if not isinstance(self.models_metrics, pd.DataFrame):
+            raise ValueError("Incorrect input type. `metrics_data` should be a DataFrame")
+        if Columns.Split not in self.models_metrics.columns:
+            raise KeyError("Missing `Split` column in `metrics_data` DataFrame")
+        if Columns.Model not in self.models_metrics.columns:
+            raise KeyError("Missing `Model`` column in `metrics_data` DataFrame")
+        if len(self.models_metrics.columns) < 3:
+            raise KeyError("`metrics_data` DataFrame assumed to have at least one metric column")
+
+    def _make_chart_data(self, fold_number: int) -> pd.DataFrame:
+        return self.models_metrics[self.models_metrics[Columns.Split] == fold_number].drop(columns=Columns.Split)
+
+    def _make_chart_data_avg(self) -> pd.DataFrame:
+        return self.models_metrics.drop(columns=Columns.Split).groupby(Columns.Model, sort=False).mean().reset_index()
+
+    def _create_chart(
+        self,
+        fig_widget: go.FigureWidget,
+        metric_x: widgets.Dropdown,
+        metric_y: widgets.Dropdown,
+        use_avg: widgets.Checkbox,
+        fold_i: widgets.Dropdown,
+    ) -> None:
+        data = self._make_chart_data_avg() if use_avg.value else self._make_chart_data(fold_i.value)
+        with fig_widget.batch_update():
+            fig_widget.data = []
+            for model in self.model_names:
+                df = data[data[Columns.Model] == model]
+                template = (f"<b>{model}</b><br>{metric_x.value}: %{{x}}<br>{metric_y.value}: %{{y}}<extra></extra>",)
+                fig_widget.add_scatter(
+                    x=df[metric_x.value],
+                    y=df[metric_y.value],
+                    mode="markers",
+                    name=model,
+                    hovertemplate=template,
+                    showlegend=self.show_legend,
+                )
+            fig_widget.layout.xaxis.title = metric_x.value
+            fig_widget.layout.yaxis.title = metric_y.value
+        self.fig = go.Figure(fig_widget.data)
+
+    def _update_chart(
```

Comment:
this method looks very very similar to the `_create_chart`. It'd be nice to avoid the code duplication

Reply:
Done. Initially I try to keep create and update steps separately from each other, but yes, after all chart tuning functions used to look quite similar.